graphblas_algorithms/algorithms/shortest_paths/dense.py

```diff
@@ -0,0 +1,47 @@
+from graphblas import Matrix, Vector, binary
+from graphblas.select import offdiag
+from graphblas.semiring import any_plus
+
+__all__ = ["floyd_warshall"]
+
+
+def floyd_warshall(G, is_weighted=False):
+    # By using `offdiag` instead of `G._A`, we ensure that D will not become dense.
+    # Dense D may be better at times, but not including the diagonal will result in less work.
+    # Typically, Floyd-Warshall algorithms sets the diagonal of D to 0 at the beginning.
+    # This is unnecessary with sparse matrices, and we set the diagonal to 0 at the end.
+    # We also don't iterate over index `i` if either row i or column i are empty.
+    if G.is_directed():
+        A, row_degrees, column_degrees = G.get_properties("offdiag row_degrees- column_degrees-")
+        nonempty_nodes = binary.pair(row_degrees & column_degrees).new(name="nonempty_nodes")
+    else:
+        A, nonempty_nodes = G.get_properties("offdiag degrees-")
```

Comment:
Note that we use some shorthand notation here. `"degrees-"` _does not_ include self-edges (i.e., diagonals), but `"degrees+"` _does_ include self-edges.